indexer/editable_map_object.cpp

```diff
@@ -49,14 +97,53 @@ vector<feature::Metadata::EType> const & EditableMapObject::GetEditableFields()
 
 StringUtf8Multilang const & EditableMapObject::GetName() const { return m_name; }
 
-vector<LocalizedName> EditableMapObject::GetLocalizedNames() const
+NamesDataSource EditableMapObject::GetNamesDataSource() const
 {
-  vector<LocalizedName> result;
-  m_name.ForEach([&result](int8_t code, string const & name) -> bool
-                 {
-                   result.push_back({code, name});
-                   return true;
-                 });
+  const auto mwmInfo = GetID().m_mwmId.GetInfo();
+
+  if (!mwmInfo)
+    return NamesDataSource();
+
+  vector<int8_t> mwmLanguages;
+  mwmInfo->GetRegionData().GetLanguages(mwmLanguages);
+
+  return GetNamesDataSource(m_name, mwmLanguages, languages::GetCurrentNorm());
+}
+
+// static
+NamesDataSource EditableMapObject::GetNamesDataSource(StringUtf8Multilang const & source,
+                                                      vector<int8_t> const & nativeMwmLanguages,
+                                                      string const & userLanguage)
+{
+  NamesDataSource result;
```

Comment:
The code below is complex. All what you need is to extract all names from source, and then sort them by languages, where languages are ordered in the following way:
* mandatory languages are less than user language
* user language is less than all other languages
```